Firestore/core/src/firebase/firestore/remote/serializer.cc

```diff
@@ -1492,101 +1493,202 @@ google_firestore_v1_MapValue Serializer::EncodeMapValue(
   return result;
 }
 
-// TODO(varconst): everything below
-
 MutationResult Serializer::DecodeMutationResult(
+    nanopb::Reader* reader,
     const google_firestore_v1_WriteResult& write_result,
     const SnapshotVersion& commit_version) const {
-  // TODO(varconst): implement.
-  // - (MutationResult)decodedMutationResult:(GCFSWriteResult *)mutation
-  //     commitVersion:(const SnapshotVersion &)commitVersion;
-  return MutationResult{{}, {}};
+  // NOTE: Deletes don't have an update_time, use commit_version instead.
+  SnapshotVersion version =
+      write_result.has_update_time
+          ? DecodeSnapshotVersion(reader, write_result.update_time)
+          : commit_version;
+  absl::optional<std::vector<FieldValue>> transform_results;
+  if (write_result.transform_results_count > 0) {
+    transform_results = std::vector<FieldValue>{};
+    for (pb_size_t i = 0; i < write_result.transform_results_count; i++) {
+      transform_results->push_back(
+          DecodeFieldValue(reader, write_result.transform_results[i]));
+    }
+  }
+  return MutationResult(std::move(version), std::move(transform_results));
 }
 
 std::unordered_map<std::string, std::string>
 Serializer::EncodeListenRequestLabels(const QueryData& query_data) const {
-  // TODO(varconst): implement.
-  // - (nullable NSMutableDictionary<NSString *, NSString *> *)
-  //     encodedListenRequestLabelsForQueryData: (const QueryData &)queryData;
-  return {};
+  auto value = EncodeLabel(query_data.purpose());
+
+  std::unordered_map<std::string, std::string> result(1);
```

Comment:
Note: this will always allocate a single entry in the map, even if `value` is empty. How about:
```cpp
auto value = EncodeLabel(query_data.purpose());
if (value.empty()) {
  return {};
}

// ...Like before, but without checking for `empty` again...
```

Also, the Objective-C code preallocates capacity, whereas the current implementation creates a map with a single "blank" element. Calling `reserve` would be more similar, but I don't think it matters at all here, so I'd suggest just default-constructing the map.